### PR TITLE
Magento 2 Zero downtime deployment

### PR DESCRIPTION
**Pull request by unailopez16**
- [ ] Bug fix
- [X] New feature  #2202
- [ ] BC breaks
- [ ] Deprecations
- [ ] Tests added
- [ ] Changelog updated

Implements check for Magento 2 database migration needs, in order to get zero downtime deployments otherwise.

Also fixes magento 2.4 deployments due to composer autoload issue https://github.com/magento/magento2/issues/23251

Tested with Deployer 7-beta.3 and Magento 2.2 | 2.3 | 2.4 (to date only 2.3 and 2.4 are officially supported by magento).

I've checked it does not introduce so far BC breaks for these versions.

### CODE REVIEW

File: recipe/magento2.php [hunk 1]
```diff
@@ -19,15 +25,15 @@
...
-    'var/cache',
     'var/export',
     'var/report',
+    'var/import',
     'var/import_history',
     'var/session',
     'var/importexport',
     'var/backups',
     'var/tmp',
-    'pub/sitemaps',
```
Comment (peterjaap) [inline recipe/magento2.php:30]: Why did you change this?

File: recipe/magento2.php [hunk 2]
```diff
@@ -48,13 +54,14 @@
 // Tasks
 desc('Compile magento di');
 task('magento:compile', function () {
+    run('cd {{release_path}} && {{bin/composer}} dump-autoload -o');
```
Comment (peterjaap) [inline recipe/magento2.php:57]: Any specific reason why this is in here?

File: recipe/magento2.php [hunk 3]
```diff
@@ -48,13 +54,14 @@
...
     run("{{bin/php}} {{release_path}}/bin/magento setup:di:compile");
     run('cd {{release_path}} && {{bin/composer}} dump-autoload -o');
 });
 
 desc('Deploy assets');
 task('magento:deploy:assets', function () {
-    run("{{bin/php}} {{release_path}}/bin/magento setup:static-content:deploy {{static_content_locales}}");
+    run("{{bin/php}} {{release_path}}/bin/magento setup:static-content:deploy {{static_content_locales}} --force");
```
Comment (peterjaap) [inline recipe/magento2.php:64]: Any specific reason why we're suddenly forcing this?

File: recipe/magento2.php [hunk 4]
```diff
@@ -69,7 +76,39 @@
...
+        if ($e->getExitCode() == DB_UPDATE_NEEDED_EXIT_CODE) {
+            $databaseUpgradeNeeded = true;
+        } else {
+            throw $e;
+        }
+    }
+
+    if ($databaseUpgradeNeeded) {
+        // check whether maintenance mode is enabled or not in order to keep the status afterwards
+        $maintenanceModeStatusOutput = run("{{bin/php}} {{release_path}}/bin/magento maintenance:status");
```
Comment (peterjaap) [inline recipe/magento2.php:99]: Wouldn't it be quicker to check for the maintenance flag instead of kicking off another bin/magento command?

File: recipe/magento2.php (hunk 3 above)
Comment (unailopez16) [reply recipe/magento2.php:64]: Starting magento 2.2 i guess, static content deploy throws an error when env is not in production mode
I may have a reason to have my environment in developer o default mode but still want to deploy it  
--force just allows to generate the static content in any mode

File: recipe/magento2.php (hunk 4 above)
Comment (unailopez16) [reply recipe/magento2.php:99]: Indeed it would be, but I decided to stick to the framework tools.  Do you see any other cons beside speed?

File: recipe/magento2.php (hunk 1 above)
Comment (peterjaap) [reply recipe/magento2.php:30]: I think we should add 'pub/sitemap' (instead of replace) to keep backwards compatibility, or we should include this change in the next major release.

File: recipe/magento2.php (hunk 2 above)
Comment (peterjaap) [reply recipe/magento2.php:57]: This problem should not occur on deploy - the generated folder will be recreated each deploy, leaving no stale classes behind.

File: recipe/magento2.php (hunk 3 above)
Comment (peterjaap) [reply recipe/magento2.php:64]: I think your case is an exception - by default, Magento should not be in developer mode (and never in default mode - that is the worst of both worlds) when deploying to production. Leaving `--force` would generate an error - exactly as it should. If you have an exception use case, you can override this in your own customizations to the recipe.

File: recipe/magento2.php (hunk 4 above)
Comment (peterjaap) [reply recipe/magento2.php:99]: Not really, this is fine, just wanted to know the reason :)

File: recipe/magento2.php (hunk 1 above)
Comment (peterjaap) [reply recipe/magento2.php:30]: It doesn't break BC in terms of deployment, but people's sitemaps could be gone if they've relied on the pub/sitemaps naming dictated by this recipe.

Comment (antonmedv) [reply recipe/magento2.php:30]: We are actually about to land v7. So breaking BC is ok.

Comment (peterjaap) [reply recipe/magento2.php:30]: Fine by me!

File: recipe/magento2.php (hunk 2 above)
Comment (unailopezvdshop) [reply recipe/magento2.php:57]: was impossible for me to deploy a m2.4 without dumping the autoload before di-compiling, got this error all the time
```
Warning: include(/var/www/html/vendor/composer/../../generated/code/Magento/Framework/App/ResourceConnection/Proxy.php):
failed to open stream: No such file or directory in /var/www/html/vendor/composer/ClassLoader.php on line 444
```
everything fine in m2.2 and m2.3

File: recipe/magento2.php (hunk 3 above)
Comment (unailopezvdshop) [reply recipe/magento2.php:64]: i agree with the production envs part, but in a normal workflow we also use deployer for dev and test environments, were it's usual  to have them in developer mode in order to get more insights about errors for instance  
do you see any way to compatibilize this at least for previous environments?

Comment (peterjaap) [reply recipe/magento2.php:64]: Use `on()`? https://deployer.org/docs/api.html#on

File: recipe/magento2.php (hunk 2 above)
Comment (peterjaap) [reply recipe/magento2.php:57]: Ok LGTM

File: recipe/magento2.php (hunk 3 above)
Comment (unailopezvdshop) [reply recipe/magento2.php:64]: don't see the way to identify previous envs, since there could be a wide variety on stage naming conventions: dev, development, testing, staging...  
anyway thinking twice we should not leave responsability of checking appropiate env mode to this static-content command, but maybe moving this check to an independent task when we find the way to identify non-production environments

Comment (peterjaap) [reply recipe/magento2.php:64]: I'm still in favor of not putting this in the default recipe at all and leaving it up to the user of the recipe to think of a way to identify their non-production environments by overriding this task, if they want to.

Comment (unailopez16) [reply recipe/magento2.php:64]: fine for me

Comment (barryvdh) [reply recipe/magento2.php]: I would skip the --force indeed.

**Comment by peterjaap** [comment]
LGTM!

**Comment by barryvdh** [comment]
Does this check also take care of config.php updates? I think Deployer Plus checks that as well? (Is normally done by the setup upgrade, right?)

**Comment by unailopez16** [comment]
The new feature keeps using the same `setup:upgrade` command, no changes on that.  
It handles the config.php file by adding new added modules to it in case the file is not control versioned.  
Ideal case is to control version the config.php, in that case setup:upgrade command won't add anything to it since changes already come from the new release git clone.  
  
Regarding other configurations stored in config.php (themes, stores, general config...), current recipe still does not have anything implemented to automatically import this new config (e.g app:config:import), could be made in the scope of another PR

**Comment by barryvdh** [comment]
If I make a change locally to app/etc/config.php (which is in git control) I get this warning:
> Exception #0 (Magento\Framework\Exception\LocalizedException): The configuration file has changed. Run the "app:config:import" or the "setup:upgrade" command to synchronize the configuration.

So that would suggest that config changes are handled by the setup:upgrade command as well. I'm just not sure if config changes are picked up by that db:status command, or that this check is also needed: https://github.com/jalogut/magento2-deployer-plus/blob/master/recipe/magento_2_2_5/config.php Otherwise the config would not be imported anymore when not database changes are made.

**Comment by antonmedv** [comment]
Some test failing?

**Comment by unailopez16** [comment]
@barryvdh you are totally right, new config (system, scopes...) would not be imported unless database changes forced a `setup:upgrade`.  

Looks like we need to add a `app:config:import`  task in order to handle this use case.

**Comment by unailopez16** [comment]
> Looks like we need to add a `app:config:import` task in order to handle this use case.

Have a look please https://github.com/deployphp/deployer/pull/2262/commits/cbe59376a73c6a2ef122dd2ee946b365fbba1167

This implementation approach also uses `app:config:status` command to detect changes in configuration, which was introduced in magento 2.2.4 (Apr '18, currently out of support by magento). Lower version deploys would fail.

**Comment by barryvdh** [comment]
So this would be breaking, should we rename this to magento2.3 or something? 
Also the code maintenance code is duplicated, but if you do setup first, it wouldn't need the config anymore, right?

**Comment by unailopez16** [comment]
> So this would be breaking, should we rename this to magento2.3 or something?

other option could be maintaining compatibility checking the current version and taking this approach only with newer versions. What do you think?


> Also the code maintenance code is duplicated, but if you do setup first, it wouldn't need the config anymore, right?

I thought about the code maintenance duplicity, but it's a tricky one. Could happen that config:import command is needed but not  setup:upgrade. I guess both commands need a maintenance mode active to be safely executed.

File: recipe/magento2.php [hunk 5]
```diff
@@ -67,9 +75,78 @@
     run("if [ -d $(echo {{current_path}}) ]; then {{bin/php}} {{current_path}}/bin/magento maintenance:disable; fi");
 });
 
+desc('Config Import');
+task('magento:config:import', function () {
+    $configImportNeeded = false;
+
+    try {
+        run('{{bin/php}} {{release_path}}/bin/magento app:config:status');
+    } catch (ProcessFailedException $e) {
```
Comment (antonmedv) [inline recipe/magento2.php:84]: This already not needed. v7 throws RunExection for localhosts too now.

File: recipe/magento2.php [hunk 6]
```diff
@@ -67,9 +75,78 @@
...
+});
+
 desc('Upgrade magento database');
 task('magento:upgrade:db', function () {
-    run("{{bin/php}} {{release_path}}/bin/magento setup:upgrade --keep-generated");
+    $databaseUpgradeNeeded = false;
+
+    try {
+        run('{{bin/php}} {{release_path}}/bin/magento setup:db:status');
+    } catch (ProcessFailedException $e) {
```
Comment (antonmedv) [inline recipe/magento2.php:121]: Same here.

**Comment by antonmedv** [comment]
So, thus pr ready to merge?

**Comment by unailopez16** [comment]
still pending to decide if it would make sense to keep compatibility with magento <= 2.2.3
what are your thoughts?

**Comment by antonmedv** [comment]
Let's introduce `magento_version` config:
```php
set('magento_version', function () {
    // detect version
    return run('{{bin/magento}} --version');
});
```

And add if (com_version) there needed. For example take a look at laravel recipe.

**Review by antonmedv — APPROVED** [review]
